gradio-app/gradio
# make all component SSR compatible

**description**

closes #9111 closes #9112

Makes all component SSR compatible.

- Removes any bare references to browser only APIs, or hide them behind on Mount hooks or actions we know can only occur when we are in the browser and fully loaded
- Stop using location.origin and instead use the root url value that we have access to.
- Added some sensible defaults to avoid things jumping around too much when hydrated in the browser
- Use empty value for certain components until we are in the browser, things like native plots or model 3d don't make much sense in a server context. We can improve this later if needed.
- Added an SSR-able test app with some example values.
- Switch to an isomophic (server + browser compatible) version of `dompurify`.



**to test**

```
pnpm i
cd js/component-test
pnpm dev
```

While running the app. I would recommend doping the following to see how things behave:

- Visit every link at the top and hard refresh the browser (with dev tools closed to you get cached fonts). This is Server Side Rendering + Client Side Rendering (hydration).
- Try with JS disabled:
   - Go into chrome dev tools
   - Click the cog in the top right 
   - scroll down to "Debugger"
   - Click "Disable JavaScript"
   - Keep dev tools open and revisit all links + refresh, this is SSR only.
